src/librbd/ImageState.cc

```diff
@@ -23,6 +24,155 @@ using util::create_async_context_callback;
+  }
+
+  ~ImageUpdateWatchers() {
+    assert(m_watchers.empty());
+    assert(m_in_flight.empty());
+    assert(m_pending_unregister.empty());
+  }
+
+  void shut_down(Context *on_finish) {
+    CephContext *cct = m_image_ctx->cct;
+    ldout(cct, 20) << __func__ << dendl;
+    {
+      Mutex::Locker locker(m_lock);
+      assert(m_on_shut_down_finish == nullptr);
+      m_watchers.clear();
+      if (!m_in_flight.empty()) {
+	m_on_shut_down_finish = on_finish;
+	return;
+      }
+    }
+    ldout(cct, 20) << __func__ << ": completing shut down" << dendl;
+    on_finish->complete(0);
+  }
+
+  void register_watcher(UpdateWatchCtx *watcher, uint64_t *handle) {
+    CephContext *cct = m_image_ctx->cct;
+    ldout(cct, 20) << __func__ << ": watcher=" << watcher << dendl;
+
+    Mutex::Locker locker(m_lock);
+    assert(m_on_shut_down_finish == nullptr);
+
+    *handle = next_handle++;
+    m_watchers.insert(std::make_pair(*handle, watcher));
+  }
+
+  void unregister_watcher(uint64_t handle, Context *on_finish) {
+    CephContext *cct = m_image_ctx->cct;
+    ldout(cct, 20) << __func__ << ": handle=" << handle << dendl;
+
+    int r = 0;
+    {
+      Mutex::Locker locker(m_lock);
+      auto it = m_watchers.find(handle);
+      if (it == m_watchers.end()) {
+	r = -ENOENT;
+      } else {
+	if (m_in_flight.find(handle) != m_in_flight.end()) {
+	  assert(m_pending_unregister.find(handle) == m_pending_unregister.end());
+	  m_pending_unregister[handle] = on_finish;
+	  on_finish = nullptr;
+	}
+	m_watchers.erase(it);
+      }
+    }
+
+    if (on_finish) {
+      ldout(cct, 20) << __func__ << ": completing unregister" << dendl;
+      on_finish->complete(r);
+    }
+  }
+
+  void notify() {
+    CephContext *cct = m_image_ctx->cct;
+    ldout(cct, 20) << __func__ << dendl;
+
+    Mutex::Locker locker(m_lock);
+    for (auto it : m_watchers) {
+      send_notify(it.first, it.second);
+    }
+  }
+
+  void send_notify(uint64_t handle, UpdateWatchCtx *watcher) {
+    assert(m_lock.is_locked());
+
+    CephContext *cct = m_image_ctx->cct;
+    ldout(cct, 20) << __func__ << ": handle=" << handle << ", watcher="
+		   << watcher << dendl;
+
+    m_in_flight.insert(handle);
+
+    Context *ctx = new FunctionContext(
+      [this, handle, watcher](int r) {
+	handle_notify(handle, watcher);
+      });
+
+    m_image_ctx->op_work_queue->queue(ctx, 0);
+  }
+
+  void handle_notify(uint64_t handle, UpdateWatchCtx *watcher) {
+
+    CephContext *cct = m_image_ctx->cct;
+    ldout(cct, 20) << __func__ << ": handle=" << handle << ", watcher="
+		   << watcher << dendl;
+
+    watcher->handle_notify();
+
+    Context *on_unregister_finish = nullptr;
+    Context *on_shut_down_finish = nullptr;
+
+    {
+      Mutex::Locker locker(m_lock);
+
+      auto in_flight_it = m_in_flight.find(handle);
+      assert(in_flight_it != m_in_flight.end());
+      m_in_flight.erase(in_flight_it);
+
+      if (m_in_flight.find(handle) == m_in_flight.end()) {
```

Comment:
Yes. m_in_flight is a mutliset, so it queues notifications with the same handle.
I can use std::map<uint64_t, int> (handle,count) instead if you think it is more readable.